osu.Game.Tournament/Models/ChessPlacement.cs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Newtonsoft.Json;
+
+namespace osu.Game.Tournament.Models
+{
+    /// <summary>
+    /// Represents a placement of a chess piece.
+    /// </summary>
+    [Serializable]
+    public class ChessPlacement
+    {
+        /// <summary>
+        /// The team holding the chess right now, representing by a <see cref="TeamColour"/>.
+        /// </summary>
+        /// <remarks>This property isn't supposed to set directly. Use <see cref="Update"/> and <see cref="Undo"/> instead.</remarks>
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate)]
+        public TeamColour OwnerTeam { get; private set; }
+
+        /// <summary>
+        /// The current <see cref="ChoiceType"/> of the chess.
+        /// </summary>
+        /// <remarks>This property isn't supposed to set directly. Use <see cref="Update"/> and <see cref="Undo"/> instead.</remarks>
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate)]
+        public ChoiceType CurrentType { get; private set; }
+
+        /// <summary>
+        /// The ID of the beatmap the chess is associated with. Zero if it isn't associated with one.
+        /// </summary>
+        public readonly int BeatmapID;
+
+        public readonly int BoardRow;
+        public readonly int BoardColumn;
+
+        /// <summary>
+        /// A list of changes of this chess piece for easier backtracking.
+        /// </summary>
+        public readonly List<ChessRecord> Records = new List<ChessRecord>();
+
+        public ChessPlacement(int boardRow, int boardColumn, TeamColour ownerTeam = TeamColour.Neutral, ChoiceType type = ChoiceType.Neutral, int beatmapID = 0)
+        {
+            BeatmapID = beatmapID;
+            OwnerTeam = ownerTeam;
+            CurrentType = type;
+
+            Debug.Assert(boardRow > 0 && boardRow <= 4, "Board row out of range.");
+            Debug.Assert(boardColumn > 0 && boardColumn <= 4, "Board column out of range.");
```

Comment:
```suggestion
            BoardRow = Math.Clamp(boardRow, 1, 4);
            BoardColumn = Math.Clamp(boardColumn, 1, 4);
```

The code can't pass `TestOutOfBoundsChess` test. It throws an exception "Board row out of range", so I reckon this indicates that the boundary checks in `ChessPlacement` haven't been aligned with what's required by the test.

This is a possible fix if you would like to have one: